### PR TITLE
Cache propagating variance flags in the relationship cache

### LINKED ISSUE #32209: Inconsistent type errors depending on order of statements
**TypeScript Version:**  3.6.0-dev.20190701, 3.5.2 (Not present in 3.4.5)

**Search Terms:**

inconsistent error typing order interface 

**Code**

```ts
type A = { a: number };
type B = { b: number };
type X<T> = ({ [K in keyof T]: T[K] } & Record<string, void>)[keyof T];
type P1<T> = { data: X<T> };
type P2<T> = { data: X<T> };

interface I<T> {
    fn<K extends keyof T>(p1: P1<Pick<T, K>>, p2: P2<Pick<T, K>>): void;
}

const i: I<A & B> = null as any;
const p2: P2<A> = null as any;

// Commenting out the below line will remove the error on the `const _i: I<A> = i;`
i.fn(null as any, p2);

const _i: I<A> = i;
```

The above code sample is a simplified version of real code in our code base. I tried simplifying it further but thats as far as I got.

**Expected behavior:**

The `i.fn(null as any, p2);` should have no effect on the typing of the `const _i: I<A> = i;` line.

**Actual behavior:**

When `i.fn(null as any, p2);` is present, there is an error on `const _i: I<A> = i;`, 
When `i.fn(null as any, p2);` is removed OR moved after `const _i: I<A> = i;`, the error disappears.

Error:
```
test.ts:15:7 - error TS2322: Type 'I<A & B>' is not assignable to type 'I<A>'.
  Type 'A' is not assignable to type 'A & B'.
    Property 'b' is missing in type 'A' but required in type 'B'.

15 const _i: I<A> = i;
         ~~

  test.ts:2:12
    2 type B = { b: number };
                 ~
    'b' is declared here.


Found 1 error.
```

**Playground Link:** 

I was not able to reproduce in the playground. But an empty `compilerOptions` and `tsc` will reliably reproduce the error in the listed versions.

v3.4.5 consistently provides no errors

**Related Issues:**

### PR DESCRIPTION
**Pull request by weswigham**
A sequel to #31688.

Fixes #32209

This allows `Unreliable` and `Unmeasurable` variances to propagate even for those types which are not references and do not have alias symbols. The downside is that the exact nuance of _which_ type parameter position within the type is unreliable/unmeasurable is lost (but that isn't really recoverable for non-alias-symbol/type-reference types anyway).

### CODE REVIEW

**Comment by weswigham** [comment]
The whole shebang of extra tests:
@typescript-bot test this
@typescript-bot run dt
@typescript-bot user test this
@typescript-bot perf test

**Comment by fatcerberus** [comment]
I'm getting really curious, could somebody explain to exactly me why TypeScript needs to *measure* variance?  I've never understood that.  A type parameter is either co-, contra-, in- or bivariant: I don't understand what there is to quantify.  Especially since structural typing (in theory) should mean the variance is implicit in the structure, right?

**Comment by weswigham** [comment]
> Especially since structural typing (in theory) should mean the variance is implicit in the structure, right?

It is - we measure the variance so we can avoid comparing the parts of the type we know are identical and just compare the type parameters as is appropriate for the structure they are contained within.

**Comment by weswigham** [comment]
RWC and user suite failures are with diffs identical to `master`, so every extended test suite is clean. ❤️

File: src/compiler/types.ts [hunk 1]
```diff
@@ -612,7 +612,12 @@ namespace ts {
     export const enum RelationComparisonResult {
         Succeeded = 1, // Should be truthy
         Failed = 2,
-        FailedAndReported = 3
```
Comment (ahejlsberg) [inline src/compiler/types.ts]: I'd suggest turning the three values into three distinct bit fields. Then you don't have to first & with a mask, you can just directly test the bit(s) you're interested in.

Comment (weswigham) [reply src/compiler/types.ts]: I mean, I I could also say that "Succeeded = 1 << 0", "Failed = 1 << 1", and "Reported = 1 << 2". (So "failed and reported" sets two bits)

Comment (ahejlsberg) [reply src/compiler/types.ts]: Sure, whatever works best, as long as it gets rid of the need to first mask and then compare.

Comment (weswigham) [reply src/compiler/types.ts]: @ahejlsberg done.

**Comment by WearyMonkey** [comment]
Hooray! Thank you @weswigham

**Pull request merged**